crates/notion-core/src/tool.rs

```diff
@@ -348,3 +355,94 @@ impl Tool for Yarn {
         }
     }
 }
+
+impl Tool for Npm {
+    fn new(session: &mut Session) -> Fallible<Self> {
+        session.add_event_start(ActivityKind::Npm);
+
+        let mut args = args_os();
+        let exe = arg0(&mut args)?;
+        if let Some(ref platform) = session.current_platform()? {
+            session.prepare_image(platform)?;
+
+            if platform.yarn.is_some() {
+                // If Yarn is enabled for the current project, warn the user that they shouldn't be using npm
+                style::display_warning(&String::from("This project is configured to use yarn, npm should be avoided"));
+            }
+            Ok(Self::from_components(&exe, args, &platform.path()?))
+        } else {
+            // Using 'Node' as the tool name since the npm version is derived from the Node version
+            // This way the error message will prompt the user to add 'Node' to their toolchain, instead of 'npm'
+            throw!(NoSuchToolError {
+                tool: "Node".to_string()
+            });
+        }
+    }
+
+    fn from_components(exe: &OsStr, args: ArgsOs, path_var: &OsStr) -> Self {
+        Npm(command_for(exe, args, path_var))
+    }
+
+    fn command(self) -> Command {
+        self.0
+    }
+
+    /// Perform any tasks which must be run after the tool runs but before exiting
+    fn finalize(session: &Session, maybe_status: &io::Result<ExitStatus>) {
+        if let Ok(_) = maybe_status {
+            if let Some(project) = session.project() {
+                let errors = project.autoshim();
+
+                for error in errors {
+                    display_error(&error);
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug, Fail, NotionFail)]
+#[fail(display = r#"
+'npx' is only available on Node >= 8.2.0
+
+This project is configured to use version {} of Node."#, version)]
+#[notion_fail(code = "ExecutableNotFound")]
+struct NpxNotAvailableError {
+    version: String,
+}
+
+impl Tool for Npx {
+    fn new(session: &mut Session) -> Fallible<Self> {
+        session.add_event_start(ActivityKind::Npx);
+
+        let mut args = args_os();
+        let exe = arg0(&mut args)?;
+        if let Some(ref platform) = session.current_platform()? {
+            session.prepare_image(platform)?;
+
+            // npx was only included with Node >= 8.2.0. If less than that, we should include a helpful error message
+            let required_node = VersionSpec::parse_requirements(">= 8.2.0")?;
```

Comment:
While `npx` wasn't *included* until Node 8.2.0, it's *compatible with*, and can be installed under, versions at least as far back as 4.9.1.  I don't think we should be blocking its use here on older versions where it's been installed.

Reply:
Based on the discussion in the issue #112, the `npm` (and `npx`) versions that are supported right now are just those that come bundled with `node`. There isn't a separate entry in `toolchain` for them, nor any way to install a specific version that isn't bundled with the selected version of Node.

Given that, if the selected `node` version is older than `8.2.0`, then `npx` won't be one of the binaries installed, so it won't be available on the path (unless the user has already installed it globally).

My thinking was it's better to have a nice error message that lets the user know what is going on instead of whatever generic error message is thrown when the command can't be found.

Reply:
I wouldn't expect `npx` to be pinned to a package; I'm more thinking of the `notion install npx latest` case.

Which, come to think of it, raises another question — if we have an `npx`-specific shim, should `notion install npx` be using that, or the generic shim?  Perhaps this should be tabled for now, after all.  😄

Reply:
Yeah, I believe that currently `notion install npx ...` won't work, as we only support installing `node` and `yarn`. This isn't changing that model at all, just supporting the fact that `npm` and `npx` are binaries that are available when you install `node`.

I agree when we have a more robust implementation that allows directly managing the versions of `npm` (and presumably `npx`), then we should remove this block, but for the current approach where we know ahead of time it's going to fail, may as well be nice about it 😄

Reply:
I've created #206 to track the question.